### ISSUE TITLE
BinaryFormatter Deserialize Datatable throw "Type * is not deserializable." Exception 

### ISSUE THREAD

**Issue by DarkGraySun** [dotnet/runtime#35346]
I have an application in .NetFramework (4.7) that serializes a datatable binary to a file. When trying to deserialize in .Net Core 3.1 the exception is thrown "Type 'System.String' is not deserializable."

Here is the sample code:
Serialization in .NetFramework 4.7:
```
 static void Main(string[] args)
        {
            var TestTable = new DataTable("TestTable");
            TestTable.Columns.Add(new DataColumn() { DataType = typeof(string), ColumnName = "TestColumn" });
            TestTable.RemotingFormat = SerializationFormat.Binary;
            var r = TestTable.NewRow();
            r["TestColumn"] = "Test";
            TestTable.Rows.Add(r);

            System.Runtime.Serialization.Formatters.Binary.BinaryFormatter binFormat = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
            using (var memstream=new System.IO.MemoryStream())
            {
                binFormat.Serialize(memstream, TestTable);
                System.IO.File.WriteAllBytes("TestTableSerialized", memstream.ToArray());
            }
        }
```

Deserialization in .Net Core 3.1:
```
static void Main(string[] args)
        {
            using (var memstream = new System.IO.MemoryStream(System.IO.File.ReadAllBytes("TestTableSerialized")))
            {
                System.Runtime.Serialization.Formatters.Binary.BinaryFormatter binFormat = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
                var table = (DataTable)binFormat.Deserialize(memstream);
            }
        }
```

**Comment by danmosemsft** [comment]
This seems to mean that a string is getting serialized as a `UnityType`. Those are for types that never have more than one instance. We're not expecting that except for `DBNull`. It would need debugging on the .NET Framework side to understand why.

**Comment by danmosemsft** [comment]
Does changing to `TestTable.RemotingFormat = SerializationFormat.Xml;` work for you? That uses a completely different mechanism so it won't hit this problem.

It is possible, but unlikely I think, that this is a duplicate of https://github.com/dotnet/runtime/issues/31611

**Comment by GrabYourPitchforks** [comment]
It means that somebody tried to deserialize `typeof(string)` (that is, the `System.Type` corresponding to _System.String_). In .NET Core, we don't allow `Type` instances to be deserialized via `BinaryFormatter`. Since presumably `DataTable` relies on this I wonder if using _SerializationFormat.Binary_ is supported in general in .NET Core?

**Comment by danmosemsft** [comment]
Ah - thank you @GrabYourPitchforks . Yes indeed:

https://referencesource.microsoft.com/#System.Data/fx/src/data/System/Data/DataTable.cs,363
`info.AddValue(String.Format(formatProvider, "DataTable.DataColumn_{0}.DataType", i), Columns[i].DataType);`

This means that you're correct, we cannot support `SerializationFormat.Binary`. And this issue is won't-fix.

Also given this
https://referencesource.microsoft.com/#System.Data/fx/src/data/System/Data/DataTable.cs,238

we probably can won't fix #31611 since we will never encounter `SerializationFormat.Xml` in the blob. Do you agree?

**Comment by GrabYourPitchforks** [comment]
I think we still need to consider fixing https://github.com/dotnet/runtime/issues/31611. That issue discusses using .NET Core to serialize a `DataSet`, then passing that payload to Full Framework for deserialization. The fix should be a one-liner: slap a `[TypeForwardedFrom]` attribute on the `SerializationFormat` enum definition.

**Comment by danmosemsft** [comment]
Ah, of course. Well, we clearly have to close this one. @DarkGraySun hopefully you can use Xml serialization, or some other approach; .NET Core does not support binary serializing Type objects.

**Comment by DarkGraySun** [comment]
Thank you for your prompt reply.
Unfortunately XML doesn't work for me.
1. It has a significantly higher resource consumption (RAM). In a test with 2 million rows, the consumption was 500 MB in binary mode and 2 GB in XML mode.

2. The DotNetFramework software with binary mode has been used for over 5 years. I cannot change the software.

**Comment by danmosemsft** [comment]
1. Can you share your scenario? Is this a matter of Core on server and Framework on app?
2. Also, do you require serialization to go in the opposite direction as well?
3. What are the data types of your columns? I'm wondering whther they are all simple types like string and int.

**Comment by DarkGraySun** [comment]
1.The Szeanrio: There is a 10 year old DotNetFramework Server app that sends compressed binary serialized data tables to the client apps. The client apps use this data as the basis for their calculations.

I now wanted to build a new client app with DotNetCore without changing the existing client apps and server apps. (Therefore I would like to avoid switching to XML.)

2. At the moment I only need the DotNetFramework to DotNetCore direction.

3. The data includes article data, customer data and price data so columns with int32, int64, string, decimal date etc. are required.

**Comment by danmosemsft** [comment]
@DarkGraySun thanks. The problem we have is that binary serialization done with BinaryFormatter (what is used here) is intrinsically fragile and prone to security bugs. We have brought it forward to .NET Core, but limited the types it supports to a relatively small list that we can control more easily. `Type` is not and cannot be on that list. It's unfortunate if the column types are actually binary serializable (eg., int, string) because if the implementation had simply serialized their names, it's possible something could have been done. That's why I was curious, although it doesn't help solve this.

Unfortunately I think your options are either change the existing client/server apps (eg., to XML) or use .NET Framework in this case. If payload size is an issue -- perhaps it could be compressed eg with `GZipStream` before transmission.

**Comment by GrabYourPitchforks** [comment]
If the case we're talking about is a .NET Framework server sending data to a .NET Core client, and if the rows contain primitive data types, this is something that might have an immediate workaround by using `BinaryReader` and `BinaryWriter`. For example (pseudo-code):

```cs
// serializing
public void Serialize(DataTable myDataTable, BinaryWriter writer)
{
    foreach (DataRow row in myDataTable.Rows)
    {
        writer.Write((string)row["customerName"]);
        writer.Write((int)row["itemQuantity"]);
        /* etc. */
    }
}
```

For types like `decimal` you could use `decimal.ToString("C", CultureInfo.InvariantCulture)` or similar to turn it into something that the writer can understand. `decimal.Parse` can reverse the transformation on the reading side.

The `BinaryWriter` format is fairly compact. It can also be further run through a compression stream if you want to gzip it. You can read the data back by using `BinaryReader` on the client.

This would involve a code change on the server to support this new endpoint, so there is some work involved. But it should be a way to immediately unblock the ability to make forward progress. It also removes the dependency on the fragile / unsafe `BinaryFormatter` code paths as Dan mentioned earlier.

**Comment by DarkGraySun** [comment]
Thank you for your effort. I was hoping that "BinaryFormatter.SurrogateSelector" or "BinaryFormatter.Binder" would offer a possibility, but that doesn't work. I will now add a new endpoint to the server to work with compressed XML. Final when the server is switched to DotNetCore I will switch back to binary.
